odoo/cli/obfuscate.py

```diff
@@ -0,0 +1,249 @@
+                        ]
+
+                    if opt.fields:
+                        if not opt.allfields:
+                            fields += [tuple(f.split('.')) for f in opt.fields.split(',')]
+                        else:
+                            _logger.error("--allfields option is set, ignoring --fields option")
+                    if opt.file:
+                        with open(opt.file, encoding='utf-8') as f:
+                            fields += [tuple(l.strip().split('.')) for l in f]
+                    if opt.exclude:
+                        if not opt.allfields:
+                            fields = [f for f in fields if f not in [tuple(f.split('.')) for f in opt.exclude.split(',')]]
+                        else:
+                            _logger.error("--allfields option is set, ignoring --exclude option")
+
+                    if opt.allfields:
+                        fields = self.get_all_fields()
+                    else:
+                        invalid_fields = [f for f in fields if not self.check_field(f[0], f[1])]
+                        if invalid_fields:
+                            _logger.error("Invalid fields: %s", ', '.join([f"{f[0]}.{f[1]}" for f in invalid_fields]))
+                            fields = [f for f in fields if f not in invalid_fields]
+
+                    if not opt.unobfuscate:
+                        self.confirm_not_secure()
+
+                    _logger.info("Processing fields: %s", ', '.join([f"{f[0]}.{f[1]}" for f in fields]))
+                    tables = defaultdict(set)
+
+                    for t, f in fields:
+                        if t[0:3] != 'ir_' and '.' not in t:
+                            tables[t].add(f)
+
+                    if opt.unobfuscate:
+                        _logger.info("Unobfuscating datas")
+                        for table in tables:
+                            _logger.info("Unobfuscating table %s", table)
+                            self.convert_table(table, tables[table], opt.pwd, opt.pertablecommit, True)
+
+                        if opt.vacuum:
+                            _logger.info("Vacuuming obfuscated tables")
+                            for table in tables:
+                                _logger.debug("Vacuuming table %s", table)
+                                self.cr.execute(sql.SQL("""VACUUM FULL {table}""").format(table=sql.Identifier(table)))
+                        self.clear_pwd()
+                    else:
+                        _logger.info("Obfuscating datas")
+                        self.set_pwd(opt.pwd)
+                        for table in tables:
+                            _logger.info("Obfuscating table %s", table)
+                            self.convert_table(table, tables[table], opt.pwd, opt.pertablecommit)
+
+                    self.commit()
+                else:
+                    self.rollback()
+
+        except Exception as e:  # noqa: BLE001
+            self.rollback()
```

Comment:
The cursor has been closed (by the `with ... as cr` at that point in the code.

Reply:
thanks.

Reply:
thanks